### PR TITLE
fix qr and sign edge cases

### LINKED ISSUE #942: QR decomposition -- zero assert is misleading / useless?
The assertion here that R is lower triangular and that it "has to be close to zero" doens't really add anything.
https://github.com/josdejong/mathjs/blob/master/lib/function/algebra/decomposition/qr.js#L229

In V8 and chrome, the idea of dividing by a large number to coerce floating point values to be zero is not a great check -- and the if/throw is just useless. Is there a good reason to keep this in? Why feel uneasy about coercing it to zero? Floating point is inherently going to be imprecise at times.

If you REALLY want a "close enough" check, doing an epsilon ball is good enough (especially if it can be configured).

### LINKED ISSUE #1669: math.qr returning lower triangular matrix
`math.qr([[0, 1], [1, 0]])`

throws this error: 

`Uncaught Error: math.qr(): unknown error - R is not lower triangular (element (1, 0)  = 1)`

I'm able to use the QR decomposition in other software on this example matrix.

I believe the issue is with the below line of code returning 0 when pivot is 0, so that the reflection is not properly calculated:

`const sgn = unaryMinus(sign(pivot))`

Side note - I believe the error message above should read: "R is not upper triangular."

### PR DESCRIPTION
**Pull request by harrysarson**
qr() was throwing an error if the pivots of a matrix were zero.
This commit fixes that and ensures that the results are consistant
with wolfram alpha.

Whilst doing this I found a bug in complex.js, I have added a work
around but we should flag this upstream.

I also addressed the long standing concerns about assert statements
within the body of the qr function.

Fixes #1669
Fixes #942

### CODE REVIEW

**Comment by harrysarson** [comment]
cc: @infusion regarding `complex.sign(0)`
cc: @c-dante regarding #942
cc: @kevinkelleher12  regarding #1669

**Comment by harrysarson** [comment]
I need to add a test with a complex matrix with zero pivots before we can merge :+1:

**Comment by harrysarson** [comment]
The codecov problem is not an issue - ignore that.

@josdejong I think I have fixed a bug in the `approx.equal` function but I am not sure so please take a look!

**Comment by josdejong** [comment]
Thanks Harry for picking this up! I will have a look soon, worst case coming weekend.

**Comment by josdejong** [comment]
Looks good Harry, good that you found these issues in `approx.js` too! I've done a bit of testing and comparison with WolframAlpha too. Sometimes the answers differ but are both valid (tested via `assertValidQRDecomposition`).

One small thing: can your replace with `describe.only('qr', ...)` with `describe('qr', ...)` again :grin: ? Then we can merge.

**Comment by harrysarson** [comment]
> Sometimes the answers differ but are both valid

Could you give an example?

---

Oh dear! Well spotted I will fix!

**Comment by josdejong** [comment]
## Example 1

some of the values in the resulted Q and R are negated.

```
qr([[5, 0, 15], [223, 0, 34.5], [1, 0, 19]])
```
mathjs: https://github.com/josdejong/mathjs/blob/6d6cd4884be94174ddd8af555d0271320b4cb81e/test/unit-tests/function/algebra/decomposition/qr.test.js#L142-L165
wolfram:  https://www.wolframalpha.com/input/?i=qr%28%5B%5B5%2C+0%2C+15%5D%2C+%5B223%2C+0%2C+34.5%5D%2C+%5B1%2C+0%2C+19%5D%5D%29

## Example 2

Wolfram doesn't return square matrices for Q and R.

```
qr([[1,2,3], [4,5,6]])
```
mathjs: https://github.com/josdejong/mathjs/blob/6d6cd4884be94174ddd8af555d0271320b4cb81e/test/unit-tests/function/algebra/decomposition/qr.test.js#L247-L277
wolfram: https://www.wolframalpha.com/input/?i=qr%28%5B%5B1%2C+2%2C+3%5D%2C+%5B2%2C+4%2C+6%5D%2C+%5B4%2C+8%2C+9%5D%5D%29

**Comment by harrysarson** [comment]
I am happy with the differences of the two examples.

The first example shows an ambiguity inherant for unit vectors, I believe the alogrithm I went for picks the signs of the elements in `Q` so that the diagonal elements of `R` are positive.

For the second example: an `m x n` matrix `A` has a QR decomposition into `m x p` and `p x n` matrices. Now `p` must be atleast the rank of `A` but may be as higher. Wolfram choses to use the minimum value of `p` (as adding more rows to the matrix is reduntant). My alogrithm went for consistancy (size of matricies produced depends only on the size the input matrix not its values). Both are valid.

In short, thanks for testings and finding these differences! I am confident that neither example is something we should be concerned about. :)

---

I fixed the `.only` and also added a flag for mocha (`--forbid-only`) that will cause `npm test` to fail (and therefore CI) should anyone* make the same mistake again.

---

*: By anyone I largely mean me.

**Comment by josdejong** [comment]
> In short, thanks for testing and finding these differences! I am confident that neither example is something we should be concerned about. :)

Yes I agree, thanks for checking it out Harry.

I like the `--forbid-only` flag, this happens to me too every now and then :joy:

**Comment by josdejong** [comment]
This fix is now published in `v6.2.4`, thanks again!
